File: backend/app/models/record.py
from datetime import datetime, timezone
import uuid

from sqlalchemy import Integer, DateTime, Boolean, ForeignKey, CheckConstraint, UUID
from sqlalchemy.orm import Mapped, mapped_column

from dependencies.postgresql_db import Base


def utcnow():
    return datetime.now(timezone.utc)



class Record(Base):
    __tablename__ = "records"
    # Constraints
    __table_args__ = (
        CheckConstraint('points % 5 = 0', name='check_points_multiple_of_5'),
    )
    # Columns
    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), default=uuid.uuid4, unique=True, nullable=False, index=True, primary_key=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, onupdate=utcnow)
    points: Mapped[int] = mapped_column(Integer)
    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False)

    # Foreign Keys
    player_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    match_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("matches.id"), nullable=False, index=True)
    question_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("questions.id"), nullable=False, index=True)

File: packages/db/src/schema/record.ts
import {
  pgTable,
  uuid,
  varchar,
  integer,
  boolean,
  timestamp,
  check,
  index,
} from "drizzle-orm/pg-core";
import { sql } from "drizzle-orm";
import { users } from "./user";
import { matches } from "./match";
import { questions } from "./question";

/**
 * Records table — v4 schema (unchanged from v3).
 * Score records: points earned per question per player.
 */
export const records = pgTable(
  "records",
  {
    id: uuid("id").defaultRandom().primaryKey(),
    points: integer("points").notNull(),
    isDeleted: boolean("is_deleted").default(false),
    playerId: uuid("player_id")
      .notNull()
      .references(() => users.id),
    matchId: uuid("match_id")
      .notNull()
      .references(() => matches.id),
    questionId: uuid("question_id")
      .notNull()
      .references(() => questions.id),
    roundNumber: integer("round_number"),
    questionCode: varchar("question_code", { length: 25 }),
    createdAt: timestamp("created_at", { withTimezone: true }).defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true }).defaultNow(),
  },
  (t) => [
    check("check_points_multiple_of_5", sql`${t.points} % 5 = 0`),
    index("idx_records_player_id").on(t.playerId),
    index("idx_records_match_id").on(t.matchId),
    index("idx_records_question_id").on(t.questionId),
  ],
);
